mantidproject/mantid
# Refactor and move `AlgorithmRunner` to Mantid API

**Description of work.**
This PR refactors the `AlgorithmRunner` class, to create a new class called `AsyncAlgorithmRunner`, which is written without Qt and using a subscriber pattern. This means it has been moved to MantidAPI. The idea is that ALFView will make use of this class to separate the algorithm thread from the GUI thread.

The PR also moves the `QtJobRunner` and `IJobRunner` used by Reflectometry to the MantidQtWidgetsCommon library.

I have created a separate issue for refactoring the `BatchAlgorithmRunner` #35184

**To test:**
1. Open StepScan
2. Load `HYSA_2934.nxs.h5` from the system test data
3. While it is loading, it should be possible to move the step scan GUI around
4. The step scan GUI will say 'Loading...' while it is loading, and then this should disappear when loading has finished.

To test the DataSelector:
1. Load data into the Indirect Corrections interface in the Apply Absorption Corrections tab (workspace ending in `_red` can be found in the sample data set)
3. While it is loading, it should be possible to move the GUI around as algorithm execution is done on a separate thread

The `QtCatalogSearcher` in the Reflectometry GUI can be tested using the following instructions:
https://developer.mantidproject.org/Testing/ReflectometryGUI/ReflectometryGUITests.html#search-by-experiment

Part of #35114

*This does not require release notes* because **it is an internal organisational change**



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
